biahub/track.py

```diff
@@ -0,0 +1,656 @@
+            tracking_functions=tracking_functions,
+            empty_frames_idx=empty_frames_idx,
+        )
+
+    # Define path to save the tracking database and graph
+    filename = str(output_dirpath).split("/")[-1].split(".")[0]
+    databaset_path = output_dirpath.parent / f"{filename}_config_tracking" / f"{fov}"
+    os.makedirs(databaset_path, exist_ok=True)
+
+    # Perform tracking
+    click.echo("Tracking...")
+    tracking_labels, tracks_df, _ = ultrack(
+        tracking_config, foreground_mask, contour_gradient_map, yx_scale, databaset_path
+    )
+
+    # Save the tracks graph to a CSV file
+    csv_path = output_dirpath / Path(*position_key) / f"tracks_{fov}.csv"
+    tracks_df.to_csv(csv_path, index=False)
+
+    click.echo(f"Saved tracks to: {output_dirpath / Path(*position_key)}")
+
+    # Save the tracking labels
+    with open_ome_zarr(output_dirpath / Path(*position_key), mode="r+") as output_dataset:
+        output_dataset[0][:, 0, 0] = np.asarray(tracking_labels)
+
+
+@click.command()
+@input_position_dirpaths()
+@output_dirpath()
+@config_filepath()
+@sbatch_filepath()
+@local()
+@click.option(
+    "-lf_dirpaths",
+    "-l",
+    required=False,
+    default=None,
+    type=str,
+    help="Label Free Image Dirpath, if there are blanck frames in the data.",
+)
+@click.option(
+    "-segmentation_dirpaths",
+    "-s",
+    required=False,
+    default=None,
+    type=str,
+    help="Segmentation Dirpath, if there are blanck frames in the data.",
+)
+@click.option(
+    "-blank_frame_csv_path",
+    "-f",
+    required=False,
+    default=None,
+    type=str,
+    help=" Blank Frame CSV path, if there are blanck frames in the data and the csv was previous gerenrated.",
+)
+def track(
+    lf_dirpaths: str,
+    output_dirpath: str,
+    config_filepath: str,
+    input_position_dirpaths: str,
+    segmentation_dirpaths: str = None,
+    sbatch_filepath: str = None,
+    local: bool = None,
+    blank_frame_csv_path: str = None,
+) -> None:
+
+    """
+    Track nuclei and membranes in using virtual staining nuclei and membranes data.
+
+    This function applied tracking to the virtual staining data for each position in the input.
+    To use this function, install the lib as pip install -e .["track"]
+
+    Example usage:
+
+    biahub track -i virtual_staining.zarr/*/*/* -l lf_stabilize.zarr -o output.zarr -c config_tracking.yml
+
+    """
+
+    output_dirpath = Path(output_dirpath)
+
+    settings = yaml_to_model(config_filepath, TrackingSettings)
+    tracking_data = settings.tracking_config
+
+    # Create default instance
+    default_config = MainConfig()
+
+    tracking_cfg = update_model(default_config, tracking_data)
+
+    # Get the shape of the data
+    with open_ome_zarr(input_position_dirpaths[0]) as dataset:
+        T, C, Z, Y, X = dataset.data.shape
+
+    # Estimate resources
+    num_cpus, gb_ram_per_cpu = estimate_resources(shape=[T, C, Z, Y, X], ram_multiplier=16)
```

Comment:
Shouldn't ultrack's number of workers be a function of the number of CPUs per job?

And the number of parallel processes for ultrack (expect solving) scales with `T`, but this function requests resources in the order of `C x T`.